homeassistant/components/esphome/__init__.py

```diff
@@ -627,11 +627,29 @@ async def _cleanup_instance(
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload an esphome config entry."""
     entry_data = await _cleanup_instance(hass, entry)
+
+    # Persist store instance so that it's re-used when removing or loading again
+    # Otherwise race conditions could occur with two separate stores pointing to the same path
+    unloaded_stores: dict[str, Store] = hass.data.setdefault(KEY_UNLOADED_STORES, {})
+    unloaded_stores[entry.entry_id] = entry_data.store
+
     return await hass.config_entries.async_unload_platforms(
         entry, entry_data.loaded_platforms
     )
 
 
+async def async_remove_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Remove an esphome config entry."""
+    await _get_store(hass, entry).async_remove()
+
+
+def _get_store(hass: HomeAssistant, entry: ConfigEntry) -> Store:
+    return hass.data.get(KEY_UNLOADED_STORES, {}).get(
```

Comment:
We preferably keep integration data contained within a single domain key namespace.
e.g., use `hass.data[DOMAIN][UNLOADED_STORES]` instead.

Reply:
Yeah the problem is a bit that `DOMAIN` is already used for `RuntimeEntryData`. But I guess we can still use it because the keys in there are uuids, and `unloaded_stores` is not a valid uuid.

Reply:
The actual use is fairly limited:

![image](https://user-images.githubusercontent.com/195327/123774177-76535480-d8cd-11eb-8930-03e1ff0d6841.png)

So, an idea could be to add a scope those the ones in the screenshots, e.g., `hass.data[DOMAIN][DATA_RUNTIME]`

Reply:
Yeah, I went ahead and refactored it into an attr class (to have more type safety). PR is a bit bigger now, but most changes should be relatively simple.